accepted/2021/default_string_globalization.md

```diff
@@ -0,0 +1,390 @@
+     public System.String ToString(System.IFormatProvider? provider) { throw null; }
+-    public System.String ToUpper() { throw null; }
+-    public System.String ToUpper(System.Globalization.CultureInfo? culture) { throw null; }
+-    public System.String ToUpperInvariant() { throw null; }
++    public System.String ToUpperCase() { throw null; } // aliases ToUpperInvariant
+     public System.String Trim() { throw null; }
+     public System.String Trim(char trimChar) { throw null; }
+     public System.String Trim(params char[]? trimChars) { throw null; }
+     public System.String TrimEnd() { throw null; }
+     public System.String TrimEnd(char trimChar) { throw null; }
+     public System.String TrimEnd(params char[]? trimChars) { throw null; }
+     public System.String TrimStart() { throw null; }
+     public System.String TrimStart(char trimChar) { throw null; }
+     public System.String TrimStart(params char[]? trimChars) { throw null; }
+ }
+ public abstract partial class StringComparer : System.Collections.Generic.IComparer<string?>, System.Collections.Generic. IEqualityComparer<string?>, System.Collections.IComparer, System.Collections.IEqualityComparer
+ {
+     protected StringComparer() { }
+-    public static System.StringComparer CurrentCulture { get { throw null; } }
+-    public static System.StringComparer CurrentCultureIgnoreCase { get { throw null; } }
+-    public static System.StringComparer InvariantCulture { get { throw null; } }
+-    public static System.StringComparer InvariantCultureIgnoreCase { get { throw null; } }
+-    public static System.StringComparer Ordinal { get { throw null; } }
+-    public static System.StringComparer OrdinalIgnoreCase { get { throw null; } }
++    public static System.StringComparer Normal { get { throw null; } } // aliases Ordinal
++    public static System.StringComparer IgnoreCase { get { throw null; } } // aliases OrdinalIgnoreCase
+     public int Compare(object? x, object? y) { throw null; }
+     public abstract int Compare(string? x, string? y);
+-    public static System.StringComparer Create(System.Globalization.CultureInfo culture, bool ignoreCase) { throw null; }
+-    public static System.StringComparer Create(System.Globalization.CultureInfo culture, System.Globalization.CompareOptions options) { throw null; }
+     public new bool Equals(object? x, object? y) { throw null; }
+     public abstract bool Equals(string? x, string? y);
+     public static System.StringComparer FromComparison(System.StringComparison comparisonType) { throw null; }
+     public int GetHashCode(object obj) { throw null; }
+     public abstract int GetHashCode(string obj);
+ }
+ public enum StringComparison
+ {
+-    CurrentCulture = 0,
+-    CurrentCultureIgnoreCase = 1,
+-    InvariantCulture = 2,
+-    InvariantCultureIgnoreCase = 3,
+-    Ordinal = 4,
+-    OrdinalIgnoreCase = 5,
++    Normal = 4, // aliases Ordinal
++    IgnoreCase = 5, // aliases OrdinalIgnoreCase
```

Comment:
What's the replacement for `CurrentCulture` and `InvariantCulture`?

How will this impact cases like `CompareOptions.NumericOrdering`: <https://github.com/dotnet/runtime/issues/13979#issuecomment-297115553>?

How will this impact cases like `RemoveStart` and `RemoveEnd`: <https://github.com/dotnet/runtime/issues/14386>

Reply:
To use CompareInfo instead for the operation. e.g. `CultureInfo.CurrentCulture.CompareInfo.Compare(...)`. I think we need to revisit [https://github.com/dotnet/runtime/issues/13979](https://github.com/dotnet/runtime/issues/13979) or at least will make NumericOrdering is cultural independent. 

---
In reply to: [615181346](https://github.com/dotnet/designs/pull/207#discussion_r615181346) [](ancestors = 615181346)

Reply:
I don't understand why we would make this so much harder when you do want CurrentCulture{IgnoreCase}.  The concern expressed in this doc is around the defaults and the inconsistency with defaults we currently expose.  But with StringComparison/StringComparer, a developer is explicitly specifying which they want.  I don't see only downside to changing this.

Reply:
I kind with @stephentoub. Having `Normal` and `IgnoreCase` present on `StringComparer` seems sufficient as it maps to the methods. Using the enum/comparer is already the advanced mode, so hiding the other concepts feels a bit like we're trying too hard to be helpful.